core/app/models/spree/product.rb

```diff
@@ -289,6 +289,8 @@ def possible_promotions
     def total_on_hand
       if any_variants_not_track_inventory?
         Float::INFINITY
+      elsif stock_items.loaded?
+        stock_items.sum(&:count_on_hand)
       else
         stock_items.sum(:count_on_hand)
```

Comment:
I think we do not need to check if they are loaded, because ActiveRecord will load the them anyway if not. So we could just use `stock_items.sum(&:count_on_hand)` in any case and reduce the cognitive load here.

Reply:
`.sum(:count_on_hand)` (no ampersand) will do the summing in SQL without loading the records.